01-architecture.md

```diff
@@ -0,0 +1,222 @@
+
+An IPLD-based Interface Definition Language should eliminate that concern. 
+
+**Future: IPLD-based IDLs**
+
+IPLD Schemas are good for defining the structure of objects, but they do not help with expressing the interface of behaviours an object can offer (e.g. methods), inputs, and outputs.
+
+Our basic idea is to design an IPLD schema for IDLs; think of [gRPC IDLs](https://grpc.io/docs/what-is-grpc/core-concepts/), [SOAP WSDL](https://www.w3schools.com/xml/xml_wsdl.asp), etc. This schema would be standardised and versioned, and understood by the FVM SDK.
+
+This IDL would not have functions as such, but rather _labelled behaviours_. This is consistent with the actor model, where actors don't expose functions, but they are known to provide behaviours.
+
+Each behaviour would carry an IPLD Schema for the input type and the return value type, an error code enum, and potentially an identifier (which incidentally could be a method number!) to facilitate internal dispatch.
+
+Traditional "overloaded functions" can be represented as distinct labelled behaviours dispatching to the same identifier.
+
+The FVM SDK would offer utilities to pack/generate calls to another actor with their IDL.
+
+_Pseudocode:_
+
+```
+import "idls/other-actor.idl.ipld" as actor
+from sdk import calls
+
+// Res is typed automatically as Result<Type, Error>
+let res = calls.target(actor)
+               .call("behaviour")
+               .with({ Field2: Value, Field2: [Value1, Value2]})
+```
+
+**Test sandbox**
+
+The FVM project will need to ship a test actor sandbox for developers to use during their actor development process.
+
+**Foreign runtimes**
+
+Support for foreign runtimes, such as the EVM, will be introduced by emulating those runtimes over WASM by compiling the respective interpreter to WASM bytecode.
+
+We expect a performance penalty from doing this (the extent of it is to be determined), but this tradeoff is workable in exchange for the highest execution fidelity/parity.
+
+Note that adopting more performant alternatives imply compiling the respective high-level languages (e.g. Solidity) to WASM, which bears additional risks that are undesirable. Doing so would also render usesless tools that operate on their native target bytecode (e.g. auditing tools that analyse EVM bytecode, EVM bytecode inspectors, etc.) Thus we prefer to adopt an approach that allows usage of the full catalogue of preexisting tools to be preserved.
+
+Each foreign runtime will require a shim to translate the storage model, gas accounting, account/address schema, gas metering, cross-actor calls, etc. to the native FVM runtime. The shim will also handle foreign chain/VM constructs by intercepting them and adapting them to the FVM native runtime.
```

Comment:
Hm. Somewhat.

- For the EVM, yes. But that's because the EVM already has a concept of cross-actor calls and gas.
- For, e.g., the JVM, no. The default WASM gas accounting will suffice. We'd still need an SDK for cross-actor calling, but it doesn't need to translate between address schemes, etc.